# Judul e Apa

Dataset source: https://www.kaggle.com/datasets/bcakra/tournamnet-result-match-of-mpl-indonesia-season-13?select=MPL+Indonesia+Season+13+-+Hero_Role.csv

## Data Wrangling

### Loading Data

In [166]:
# Import libraries
import pandas as pd
import numpy as np

In [234]:
# Load the data
data = pd.read_csv('data/MPL Indonesia Season 13 - Hero_Role.csv')
data.head()

Hero                       Title  Hero Order    Role(s) Specialty(ies)  \
0     Miya  Miya, the Moonlight Archer           1   Marksman       Finisher   
1  Balmond   Balmond, the Bloody Beast           2    Fighter         Damage   
2    Saber  Saber, the Wandering Sword           3   Assassin         Charge   
3    Alice   Alice, the Queen of Blood           4  Mage/Tank   Charge/Regen   
4     Nana      Nana, the Sweet Leonin           5       Mage           Poke   

  Lane Recommendation(s)  Health  Health Regen   Mana  Mana Regen  ...  \
0              Gold Lane    2524           6.0  445.0        15.0  ...   
1      Jungling/EXP Lane    2736           9.4    0.0         0.0  ...   
2               Jungling    2599           7.0  443.0        16.0  ...   
3      EXP Lane/Mid Lane    2573           7.2  493.0        18.0  ...   
4       Mid Lane/Roaming    2501           6.8  510.0        18.0  ...   

                                    Pasive Deskripsi  Skill 1 Name Abilities  \
0  Miya gains 5% extra Attack Speed for 4s each t...              Moon Arrow   
1  Balmond recovers 5% Max HP on minion or creep ...               Soul Lock   
2  Saber's attacks reduce enemies' Physical Defen...         Orbiting Swords   
3  A blood orb will be generated each time an ene...           Flowing Blood   
4  Upon taking fatal damage, Nana removes all deb...         Magic Boomerang   

   Skill 1 Ability                           Skill 1 Pasive Deskripsi  \
0         BUFF AoE  Miya shoots two extra arrows on each Basic Att...   
1       Blink Slow  Balmond charges in the target direction for a ...   
2         AoE BUFF  Saber releases 5 swords that orbit around him ...   
3        Blink AoE  Alice casts a sphere of blood energy in the ta...   
4              AoE  Nana hurls her Magic Boomerang in the target d...   

   Skill 2 Name Abilities  Skill 2 Ability  \
0        Arrow of Eclipse           CC AoE   
1           Cyclone Sweep         AoE BUFF   
2                  Charge        Blink AoE   
3               Blood Awe           CC AoE   
4           Molina Smooch        CC Debuff   

                            Skill 2 Pasive Deskripsi  Skill 3 Name Abilities  \
0  Miya launches an empowered arrow on the target...        Hidden Moonlight   
1  Balmond spins with his axe for 3s, continuousl...          Lethal Counter   
2  Saber dashes in the target direction, dealing ...            Triple Sweep   
3  Alice deals 270(+90%Total Magic Power) Magic D...               Blood Ode   
4  Nana sends Molina to the target location, who ...            Molina Blitz   

     Skill 3 Ability                           Skill 3 Pasive Deskripsi  
0  Conceal Remove CC  Miya removes all debuffs on her and conceals h...  
1         Burst Slow  After a short delay, Balmond smashes his axe i...  
2           Burst CC  Saber charges at the target enemy hero, knocki...  
3           AoE HEAL  Alice enters the blood-draining state, consumi...  
4           Burst CC  Nana summons Molina to attack the target area ...  

[5 rows x 34 columns]

### Assessing Data

In [235]:
# Check the shape of the data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123 entries, 0 to 122
Data columns (total 34 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Hero                      123 non-null    object 
 1   Title                     123 non-null    object 
 2   Hero Order                123 non-null    int64  
 3   Role(s)                   123 non-null    object 
 4   Specialty(ies)            123 non-null    object 
 5   Lane Recommendation(s)    123 non-null    object 
 6   Health                    123 non-null    int64  
 7   Health Regen              123 non-null    float64
 8   Mana                      118 non-null    float64
 9   Mana Regen                118 non-null    float64
 10  Attack Speed              123 non-null    object 
 11  Physical Damage           123 non-null    int64  
 12  Magical Damage            123 non-null    int64  
 13  Physical Defense          123 non-null    int64  
 14  Magical De

Total 34 kolom dan 123 entri hero. Terdapat beberapa tipe data pada kolom krusial yang belum sesuai, seperti win rate, price, dan attack speed.

In [236]:
data.nunique()

Hero                        123
Title                       123
Hero Order                  123
Role(s)                      24
Specialty(ies)               40
Lane Recommendation(s)       18
Health                       83
Health Regen                 26
Mana                         43
Mana Regen                   19
Attack Speed                 35
Physical Damage              31
Magical Damage                1
Physical Defense             16
Magical Defense               2
Movement Speed               13
Energy                        1
Energy Regen                  5
Price                        13
Specialty                    61
Win Rate                    123
Hero.1                      123
 Name Pasive Abilities      123
Ability                      30
Pasive Deskripsi            123
Skill 1 Name Abilities      123
Skill 1 Ability              41
Skill 1 Pasive Deskripsi    123
Skill 2 Name Abilities      123
Skill 2 Ability              59
Skill 2 Pasive Deskripsi    123
Skill 3 

In [237]:
# Mengecek nilai null
data.isna().sum()

Hero                          0
Title                         0
Hero Order                    0
Role(s)                       0
Specialty(ies)                0
Lane Recommendation(s)        0
Health                        0
Health Regen                  0
Mana                          5
Mana Regen                    5
Attack Speed                  0
Physical Damage               0
Magical Damage                0
Physical Defense              0
Magical Defense               0
Movement Speed                0
Energy                      117
Energy Regen                117
Price                         1
Specialty                     3
Win Rate                      0
Hero.1                        0
 Name Pasive Abilities        0
Ability                       0
Pasive Deskripsi              0
Skill 1 Name Abilities        0
Skill 1 Ability               0
Skill 1 Pasive Deskripsi      0
Skill 2 Name Abilities        0
Skill 2 Ability               0
Skill 2 Pasive Deskripsi      0
Skill 3 

Terdapat 5 entri hero yang tidak menggunakan mana/mana regen melainkan energi/regen. 1 entri hero juga yang pricenya null. Untuk kolom specialty, dikarenakan tidak relevant dengan tujuan yg dicapai maka abaikan saja

In [238]:
data.describe()

Hero Order       Health  Health Regen        Mana  Mana Regen  \
count  123.000000   123.000000    123.000000  118.000000  118.000000   
mean    62.000000  2580.544715      7.681301  347.211864   12.708475   
std     35.651087   207.092529      1.592161  214.532322    8.567745   
min      1.000000   918.000000      3.800000    0.000000    0.000000   
25%     31.500000  2501.000000      6.800000    0.000000    0.000000   
50%     62.000000  2578.000000      7.200000  440.000000   15.000000   
75%     92.500000  2701.000000      8.000000  490.000000   18.000000   
max    123.000000  2909.000000     18.400000  750.000000   31.000000   

       Physical Damage  Magical Damage  Physical Defense  Magical Defense  \
count       123.000000           123.0        123.000000       123.000000   
mean        116.065041             0.0         19.154472        14.878049   
std           9.835316             0.0          3.550646         0.774442   
min          90.000000             0.0         10.000000        10.000000   
25%         110.000000             0.0         17.000000        15.000000   
50%         117.000000             0.0         19.000000        15.000000   
75%         122.000000             0.0         21.500000        15.000000   
max         140.000000             0.0         27.000000        15.000000   

       Movement Speed  Energy  Energy Regen  
count      123.000000     6.0      6.000000  
mean       251.495935   100.0     42.800000  
std          8.715328     0.0     96.618425  
min        240.000000   100.0      0.800000  
25%        240.000000   100.0      3.250000  
50%        252.000000   100.0      4.000000  
75%        260.000000   100.0      4.750000  
max        270.000000   100.0    240.000000

- Dikarenakan kolom magical damage bernilai 0 semua dan hero magic pun mempunyai physical damage pada tabel, anggap physical damage sebagai damage universal.
- Dikarenakan beberapa hero ada yang tidak menggunakan mana sama sekali dan beberapa hero juga menggunakan energi. Penggunaan mana,energi,mana regen, dan energi regen diabaikan saja dari kriteria.

In [239]:
print("Jumlah duplikasi: ", data.duplicated().sum())

Jumlah duplikasi:  0


### Data Cleaning

In [240]:
clean_data = data.copy()

#### Menyesuaikan tipe data

Mengubah tipe data win rate, price, dan attack speed menjadi float

In [241]:
# Mengecek tipe data win rate, price, dan attack speed
clean_data[['Win Rate', 'Price', 'Attack Speed']].dtypes

Win Rate        object
Price           object
Attack Speed    object
dtype: object

In [242]:
# Melihat entri data win rate, price, dan attack speed
clean_data[['Attack Speed', 'Price', 'Win Rate']].head()

Attack Speed        Price              Win Rate
0  1.06 ┃ 100%        10800    33W : 32L (50.77%)
1  1.06 ┃ 100%   6500   299  453W : 516L (46.75%)
2  1.08 ┃ 100%   6500   299  142W : 146L (49.31%)
3     1 ┃ 100%  15000   399  513W : 526L (49.37%)
4  1.08 ┃ 100%   6500   299   74W : 105L (41.34%)

In [243]:
# Mengganti nilai data attack speed
clean_data['Attack Speed'] = clean_data['Attack Speed'].str.split('┃').str[0].astype(float)

In [244]:
# Mengganti nilai data win rate
clean_data['Win Rate'] = clean_data['Win Rate'].str.extract(r'(\d+(\.\d+)?)%\)$')[0].astype(float)

In [245]:
# Mengganti nilai data price
clean_data['Price'] = clean_data['Price'].str.split().str[0].astype(float)

In [246]:
# Melihat entri data win rate, price, dan attack speed
clean_data[['Attack Speed', 'Price', 'Win Rate']].head()

Attack Speed    Price  Win Rate
0          1.06  10800.0     50.77
1          1.06   6500.0     46.75
2          1.08   6500.0     49.31
3          1.00  15000.0     49.37
4          1.08   6500.0     41.34

In [247]:
# Mengecek Tipe Data
clean_data[['Win Rate', 'Price', 'Attack Speed']].dtypes

Win Rate        float64
Price           float64
Attack Speed    float64
dtype: object

#### Mengisi nilai price yang hilang

In [248]:
# Melihat hero yang mempunyai price yang hilang.
clean_data[clean_data['Price'].isna()]

Hero                    Title  Hero Order Role(s) Specialty(ies)  \
118  Novaria  Novaria, the Star Rebel         119    Mage          Burst   

    Lane Recommendation(s)  Health  Health Regen   Mana  Mana Regen  ...  \
118               Mid Lane    2401           6.8  540.0         4.6  ...   

                                      Pasive Deskripsi  \
118  Astral Spheres summoned by Novaria will contin...   

     Skill 1 Name Abilities  Skill 1 Ability  \
118           Astral Meteor         Slow AoE   

                              Skill 1 Pasive Deskripsi  \
118  Novaria summons an Astral Sphere at the target...   

     Skill 2 Name Abilities  Skill 2 Ability  \
118           Astral Recall       Burst BUFF   

                              Skill 2 Pasive Deskripsi  \
118  Novaria summons an Astral Sphere at a distance...   

     Skill 3 Name Abilities  Skill 3 Ability  \
118             Astral Echo             BUFF   

                              Skill 3 Pasive Deskripsi  
118  Novaria scatters Astral Echo in the target dir...  

[1 rows x 34 columns]

Hero yang mempunyai price yang hilang adalah hero yang bernama "Novaria". Isi nilai yang hilang dengan 32000 karena harga hero novaria di game adalah 32000.

In [249]:
# Mengisi nilai
clean_data['Price'].fillna(32000, inplace=True)

In [250]:
# Mengecek nilai null pada price kembali
clean_data['Price'].isna().sum()

0

In [251]:
# Mengecek price hero novaria
clean_data[clean_data['Hero'] == 'Novaria'].Price

118    32000.0
Name: Price, dtype: float64

#### Menyeleksi Fitur yang Dibutuhkan

In [252]:
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123 entries, 0 to 122
Data columns (total 34 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Hero                      123 non-null    object 
 1   Title                     123 non-null    object 
 2   Hero Order                123 non-null    int64  
 3   Role(s)                   123 non-null    object 
 4   Specialty(ies)            123 non-null    object 
 5   Lane Recommendation(s)    123 non-null    object 
 6   Health                    123 non-null    int64  
 7   Health Regen              123 non-null    float64
 8   Mana                      118 non-null    float64
 9   Mana Regen                118 non-null    float64
 10  Attack Speed              123 non-null    float64
 11  Physical Damage           123 non-null    int64  
 12  Magical Damage            123 non-null    int64  
 13  Physical Defense          123 non-null    int64  
 14  Magical De

In [262]:
features_needed = ['Hero', 'Health', 'Health Regen', 'Attack Speed', 'Physical Damage', 'Physical Defense', 'Magical Defense', 'Movement Speed', 'Price', 'Win Rate']

In [263]:
final_data = clean_data[features_needed].copy()

In [264]:
# Rename physical damage column to damage
final_data.rename(columns={'Physical Damage': 'Damage'}, inplace=True)

In [265]:
final_data.head()

Hero  Health  Health Regen  Attack Speed  Damage  Physical Defense  \
0     Miya    2524           6.0          1.06     100                17   
1  Balmond    2736           9.4          1.06     119                25   
2    Saber    2599           7.0          1.08     118                20   
3    Alice    2573           7.2          1.00     114                21   
4     Nana    2501           6.8          1.08     115                17   

   Magical Defense  Movement Speed    Price  Win Rate  
0               15             240  10800.0     50.77  
1               15             260   6500.0     46.75  
2               15             260   6500.0     49.31  
3               15             240  15000.0     49.37  
4               15             250   6500.0     41.34

In [266]:
final_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 123 entries, 0 to 122
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Hero              123 non-null    object 
 1   Health            123 non-null    int64  
 2   Health Regen      123 non-null    float64
 3   Attack Speed      123 non-null    float64
 4   Damage            123 non-null    int64  
 5   Physical Defense  123 non-null    int64  
 6   Magical Defense   123 non-null    int64  
 7   Movement Speed    123 non-null    int64  
 8   Price             123 non-null    float64
 9   Win Rate          123 non-null    float64
dtypes: float64(4), int64(5), object(1)
memory usage: 9.7+ KB


In [267]:
final_data.describe()

Health  Health Regen  Attack Speed      Damage  Physical Defense  \
count   123.000000    123.000000    123.000000  123.000000        123.000000   
mean   2580.544715      7.681301      1.040650  116.065041         19.154472   
std     207.092529      1.592161      0.059736    9.835316          3.550646   
min     918.000000      3.800000      0.870000   90.000000         10.000000   
25%    2501.000000      6.800000      1.000000  110.000000         17.000000   
50%    2578.000000      7.200000      1.030000  117.000000         19.000000   
75%    2701.000000      8.000000      1.080000  122.000000         21.500000   
max    2909.000000     18.400000      1.220000  140.000000         27.000000   

       Magical Defense  Movement Speed         Price   Win Rate  
count       123.000000      123.000000    123.000000  123.00000  
mean         14.878049      251.495935  28222.756098   48.74813  
std           0.774442        8.715328   8120.532319    4.76978  
min          10.000000      240.000000    599.000000   25.00000  
25%          15.000000      240.000000  32000.000000   46.95500  
50%          15.000000      252.000000  32000.000000   49.37000  
75%          15.000000      260.000000  32000.000000   51.10500  
max          15.000000      270.000000  32000.000000   61.11000

In [268]:
# Simpan data yang sudah final
final_data.to_csv('./data/final_data.csv', index=False)